engine/googletest_list.cpp

```diff
@@ -0,0 +1,122 @@
+//   notice, this list of conditions and the following disclaimer in the
+//   documentation and/or other materials provided with the distribution.
+// * Neither the name of Google Inc. nor the names of its contributors
+//   may be used to endorse or promote products derived from this software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "engine/googletest_list.hpp"
+
+#include <fstream>
+#include <regex>
+#include <string>
+#include <utility>
+
+#include "engine/exceptions.hpp"
+#include "model/metadata.hpp"
+#include "model/test_case.hpp"
+#include "utils/config/exceptions.hpp"
+#include "utils/format/macros.hpp"
+
+namespace config = utils::config;
+namespace fs = utils::fs;
+
+
+namespace {
+
+/// A regular expression that should match either a test suite or a test case.
+const std::string name_expr = "([[:alpha:][:digit:]_]+[[:alpha:][:digit:]_/]*)";
+
+/// The separator between a test suite and a test case.
+const std::string testsuite_testcase_separator = ".";
+
+/// A complete regular expression representing a line with a test suite
+/// definition.
+///
+/// e.g.,
+///   "TestSuite."
+/// or
+///   "TestSuite/Prefix."
+/// or
+///   "TestSuite/Prefix.    # TypeParam = .+"
+const std::string testsuite_expr =
+    name_expr + "\\.([[:space:]]+# TypeParam = .+)?";
+
+/// A complete regular expression representing a line with a test case
+/// definition.
+///
+/// e.g.,
+///   "  TestCase"
+/// or
+///   "  TestCase/0"
+/// or
+///   "  TestCase/0  # GetParam() = 4"
+const std::string testcase_expr =
+    "  " + name_expr + "([[:space:]]+# GetParam\\(\\) = .+)?";
+
+}  // anonymous namespace
+
+
+/// Parses the googletest list of test cases from an open stream.
+///
+/// \param input The stream to read from.
+///
+/// \return The collection of parsed test cases.
+///
+/// \throw format_error If there is any problem in the input data.
+model::test_cases_map
+engine::parse_googletest_list(std::istream& input)
+{
+    std::regex testcase_re(testcase_expr), testsuite_re(testsuite_expr);
+    std::smatch match;
+    std::string line;
+    std::string testcase_name, test_suite;
```

Comment:
nit: Not C89. Move variables to the innermost scope where they are needed and merge declaration with initialization wherever possible.

Reply:
For the record, I try to avoid this in loops because it really pessimizes the program when compiled with `-O0`

Reply:
Oh, of course I didn't see this reply before adding more comments along these lines.

Why is that important though? You should expect bad performance with `-O0` and I think code readability is much more preferable than that. Sure, if it affected the optimized version it'd be a different story, but we are talking about a debug build here. The more variables that are in scope at any given time, the harder it is to understand what's going on and under what conditions (or not) those uninitialized variables may be accessed.

Reply:
I moved the variables inside the loop that could be put inside the loop. I needed to leave some variables outside the loop since they cannot be reinitialized each iteration.